### PR TITLE
Raven db 14462

### PR DESCRIPTION
**Pull request by talweiss1982**

### CODE REVIEW

File: tools/Voron.Recovery/Recovery.cs [hunk 1]
```diff
@@ -496,6 +530,24 @@ void PrintRecoveryProgress(long startOffset, byte* mem, byte* eof, DateTime now)
             }
         }
 
+        private void CheckInvalidPagesInARaw(PageHeader* pageHeader, byte* mem)
+        {
+            if(_shouldIgnoreInvalidPagesInARaw)
+                return;
+
+            if (*(short*)pageHeader->Mac != 0)
```
Comment (ayende) [inline tools/Voron.Recovery/Recovery.cs]: That isn't correct. the `Mac` is an array of 16 bytes, you need to check that the full array is set to zero

File: tools/Voron.Recovery/Recovery.cs [hunk 2]
```diff
@@ -496,6 +530,24 @@ void PrintRecoveryProgress(long startOffset, byte* mem, byte* eof, DateTime now)
...
+            {
+                if (MaxNumberOfInvalidChecksumWithNoneZeroMac <= _InvalidChecksumWithNoneZeroMac++)
+                {
+                    var error =
+                        $"Found {_InvalidChecksumWithNoneZeroMac} pages in a raw that had invalid checksum and none zero MAC " +
+                        $"this is a strong indication that you're recovering an encrypted database and didn't provide the 'Master Key'";
```
Comment (ayende) [inline tools/Voron.Recovery/Recovery.cs]:
```suggestion
                        $"this is a strong indication that you're recovering an encrypted database and didn't provide the encryption key using the  '--MasterKey=<KEY>' command line flag";
```

File: tools/Voron.Recovery/Recovery.cs [hunk 3]
```diff
@@ -910,11 +962,14 @@ private bool ValidateOverflowPage(PageHeader* pageHeader, byte* eof, long startO
...
                 if (checksum != pageHeader->Checksum)
                 {
+                    CheckInvalidPagesInARaw(pageHeader, mem);
                     var message =
                         $"Invalid checksum for overflow page {pageHeader->PageNumber}, expected hash to be {pageHeader->Checksum} but was {checksum}";
                     mem = PrintErrorAndAdvanceMem(message, mem);
                     return false;
                 }
+
+                _InvalidChecksumWithNoneZeroMac = 0;
```
Comment (ayende) [inline tools/Voron.Recovery/Recovery.cs]: If we were able to get a _single_ valid page, we don't need to deal with this any longer.

**Review by ayende — CHANGES_REQUESTED** [review]

**Comment by talweiss1982** [comment]
done

**Comment by talweiss1982** [comment]
test is failing

**Review by ayende — APPROVED** [review]

File: tools/Voron.Recovery/CommandLineApp.cs [hunk 4]
```diff
@@ -164,6 +164,15 @@ private static void ConfigureRecoveryCommand()
 
                         config.IgnoreDataIntegrityErrorsOfAlreadySyncedTransactions = ignoreDataIntegrityErrorsOfAlreadySyncedTransactions;
                     }
+                    
+                    if (ignoreInvalidPagesInARawArg.HasValue())
+                    {
+                        var value = ignoreInvalidPagesInARawArg.Value();
+                        if (bool.TryParse(value, out var ignoreInvalidPagesInARaw) == false)
+                            return ExitWithError($"{nameof(config.IgnoreDataIntegrityErrorsOfAlreadySyncedTransactions)} argument value ({value}) is invalid", cmd);
```
Comment (arekpalinski) [inline tools/Voron.Recovery/CommandLineApp.cs]: Invalid argument name in the message - `{nameof(config.IgnoreDataIntegrityErrorsOfAlreadySyncedTransactions)}`

File: test/SlowTests/Issues/RavenDB-8451.cs [hunk 5]
```diff
@@ -24,8 +24,29 @@ public RavenDB_8451(ITestOutputHelper output) : base(output)
...
+        [Fact64Bit]
+        public async Task RecoveryOfEncryptedDatabaseWithoutMasterKeyShouldThrow()
+        {
+            try
+            {
+                await CanRecoverEncryptedDatabaseInternal(true);
+            }
+            catch (InvalidOperationException ioe)
+            {
+            }
```
Comment (arekpalinski) [inline test/SlowTests/Issues/RavenDB-8451.cs]: Can we use `Assert.Throws<>()` and check the exception message to be sure that it throws what we need

File: tools/Voron.Recovery/Recovery.cs [hunk 6]
```diff
@@ -442,8 +474,13 @@ void PrintRecoveryProgress(long startOffset, byte* mem, byte* eof, DateTime now)
                                         conflictsWriter, countersWriter, context, startOffset, ((RawDataSmallPageHeader*)page)->TableType) == false)
                                     break;
                             }
+
                             mem += _pageSize;
                         }
+                        catch (InvalidOperationException ioe) when (ioe.Message.Contains("this is a strong indication that you're"))
```
Comment (arekpalinski) [inline tools/Voron.Recovery/Recovery.cs]: Can we extract error message to const and use it here?

File: tools/Voron.Recovery/Recovery.cs [hunk 7]
```diff
@@ -496,6 +533,31 @@ void PrintRecoveryProgress(long startOffset, byte* mem, byte* eof, DateTime now)
             }
         }
 
+        private void CheckInvalidPagesInARaw(PageHeader* pageHeader, byte* mem)
```
Comment (arekpalinski) [inline tools/Voron.Recovery/Recovery.cs:536]: Just to make sure.. what did you mean by "in a raw" ? I see you count it so maybe you meant "in a row" but from the other side I don't see you reset `_InvalidChecksumWithNoneZeroMac` anywhere

Comment (talweiss1982) [reply tools/Voron.Recovery/Recovery.cs:536]: It is only the first pages in a row, once a valid page checksum was found we disable this check by using the flag: _shouldIgnoreInvalidPagesInARaw

**Review by arekpalinski — APPROVED** [review]

**Pull request merged**